Move useful utility functions out into their own static class

LiruMouse opened the pull request:
DomainDice.ChangeType<T> => Utils.ChangeType<T>
DomainDice.BooleanOrType<T> => Utils.BooleanOrType<T>
Adds Utils.cs to RollGen project.

Sorry, I forgot to make this earlier.

LiruMouse committed:
Move useful utility functions out into their own static class

DomainDice.ChangeType<T> => Utils.ChangeType<T>
DomainDice.BooleanOrType<T> => Utils.BooleanOrType<T>
Adds Utils.cs to RollGen project.

cidthecoatrack commented on RollGen/Utils.cs:
```diff
@@ -0,0 +1,24 @@
+{
+    public static class Utils
+    {
+        public static T ChangeType<T>(object rawEvaluatedExpression)
```
This method could probably be private, as only the boolean seems to be used

LiruMouse replied:
@cidthecoatrack As I said earlier, people could use these utilities, so they shouldn't be private.

cidthecoatrack commented on RollGen.Domain/DomainDice.cs:
```diff
@@ -49,7 +49,7 @@ public string ReplaceWrappedExpressions<T>(string source, string expressionOpen
                 var unevaluatedMatch = Evaluate(lenientReplacedDice);
                 var target = expressionOpen + matchGroupValue + expressionClose;
-                var replacement = BooleanOrType<T>(unevaluatedMatch).ToString();
+                var replacement = Utils.BooleanOrType<T>(unevaluatedMatch).ToString();
```
Can you remind me of this context, and why we might have a boolean to return here?

cidthecoatrack replied:
Also, instead of a static class, can we make this injectable?  Like a "converter" or something.

LiruMouse replied:
@cidthecoatrack If the dice roll expression is something like 1d20 > 10 then it should return True or False, not 1 or 0.
Also, these functions are nice on their own if the end user wants to do conversion on their own from the `object`.

cidthecoatrack replied:
Do we/you have use cases where one wants to do those conversions?  Personal taste tends to dictate to make as many things private/hidden/internal as possible, and only show when there is a confirmed use case that necessitates it.  So if you already have a use case where being able to access those conversions helps, then great.  Otherwise, let's keep them hidden.

LiruMouse replied:
I do have a use case, yepyep, main reason why I had to make these accessible.

cidthecoatrack replied:
Then yes, they should stay public. :smile: I would prefer still if we made it injectable instead of static, and wrapped these functions in unit tests.  Also, we should add integration tests that verify that a boolean expression does evaluate correctly.

If we make it injectable, the interface can be public (in RollGen project), but the implementation can be internal, conforming with the standards of forcing use of Ninject to acquire implementations.

LiruMouse replied:
@cidthecoatrack That seems like a complex way to use a Utility class...

cidthecoatrack replied:
A fair note.  Well then, we can leave these as static for now, given the nature of the methods.  I do still want unit tests around them though :)

cidthecoatrack commented on RollGen/Utils.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace RollGen
+{
+    public static class Utils
```
Regardless of if this is injectable or static, we should add unit tests around this class.

cidthecoatrack replied:
Also, in line with the current solution, and given the comment above about hiding functionality, this should probably be marked internal.

LiruMouse replied:
@cidthecoatrack The reason not to is because it needs to be accessible by other projects.

LiruMouse committed:
Add Unit Tests for Utils class.

Lirusaito commented:
I added Unit tests, it's pretty hard to Unit test something becoming a boolean or a given type after its turned into a string, so I did my best on that one.
Also I had to add a function to be able to unit test generics, I based it off [this example](http://stackoverflow.com/a/2365953).

cidthecoatrack commented:
Excellent, I like those tests.  Can you add an integration test case around the use case you listed earlier, regarding `1d20 > 10`?  You might need to make some dumb cases that will have consistent answers (like `1d20 > 100` or `1d100 <= 0`), and then maybe a custom `FullRangeHit` method (verifies both `true` and `false` hit for `1d20 > 10`).

Lirusaito commented:
I don't see how this can be done in a test unless we create a new Dice that
uses Albatross as the parser (since we only parse the dice expressions in
the strings, then pass them to Albatross)

On Tue, May 24, 2016 at 9:43 AM, Karl Speer <notifications@github.com>
wrote:

> Excellent, I like those tests. Can you add an integration test case around
> the use case you listed earlier, regarding 1d20 > 10? You might need to
> make some dumb cases that will have consistent answers (like 1d20 > 100
> or 1d100 <= 0), and then maybe a custom FullRangeHit method (verifies
> both true and false hit for 1d20 > 10).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/DnDGen/RollGen/pull/27#issuecomment-221273456>
>

cidthecoatrack commented:
Here would be an example of an integration test (done as an Expression stress test)

```csharp
        [TestCase("1d20 > 10", true, false)]
        [TestCase("1d20 < 2d10", true, false)]
        [TestCase("3d2 >= 2d3", true, false)]
        [TestCase("2d6 <= 3d4", true, false)]
        [TestCase("1d2 = 2", true, false)]
        [TestCase("1d100 > 0", true)]
        [TestCase("100 < 1d20", false)]
        [TestCase("1d1 = 1", true)]
        [TestCase("9266 = 9266", true)]
        [TestCase("9266 = 90210", false)]
        [TestCase("1d2 = 3", false)]
        public void FullRangeHit(string expression, params bool[] expectedRolls)
        {
            var rolls = Populate(new HashSet<bool>(), () => GenerateBoolean(expression), expectedRolls.Length);

            Assert.That(rolls, Is.EquivalentTo(expectedRolls));
            Assert.That(expectedRolls, Is.EquivalentTo(rolls));
        }

        private bool GenerateBoolean(string expression)
        {
            var totaledExpression = Dice.ReplaceExpressionWithTotal(expression);
            var evaluatedExpression = Dice.Evaluate<bool>(totaledExpression);

            return evaluatedExpression;
        }
```

LiruMouse committed:
Add `FullRangeTruth` to ExpressionTests.cs

Makes `StressTests.Populate` a generic so it can take/generate/return collections of anything.

cidthecoatrack commented:
Awesome, thank you.  This is good to merge, and I'll release it. :smile:
